velero-csi-plugin/util.go

```diff
@@ -0,0 +1,143 @@
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get PV %s for PVC %s/%s", pvName, pvc.Namespace, pvc.Name)
+	}
+	return pv, nil
+}
+
+func getPodsUsingPVC(pvcNamespace, pvcName string, corev1 corev1client.PodsGetter) ([]corev1api.Pod, error) {
+	podsUsingPVC := []corev1api.Pod{}
+	for {
+		podList, err := corev1.Pods(pvcNamespace).List(metav1.ListOptions{})
+		if err != nil {
+			return podsUsingPVC, err
+		}
+
+		for _, p := range podList.Items {
+			for _, v := range p.Spec.Volumes {
+				if v.PersistentVolumeClaim != nil && v.PersistentVolumeClaim.ClaimName == pvcName {
+					podsUsingPVC = append(podsUsingPVC, p)
+				}
+			}
+		}
+		if podList.Continue == "" {
+			break
+		}
+	}
+
+	return podsUsingPVC, nil
+}
+
+func getPodVolumeNameForPVC(pod corev1api.Pod, pvcName string) (string, error) {
+	for _, v := range pod.Spec.Volumes {
+		if v.PersistentVolumeClaim != nil && v.PersistentVolumeClaim.ClaimName == pvcName {
+			return v.Name, nil
+		}
+	}
+	return "", errors.Errorf("Pod %s/%s does not use PVC %s/%s", pod.Namespace, pod.Name, pod.Namespace, pvcName)
+}
+
+func getPodVolumesUsingRestic(pod corev1api.Pod) []string {
+	resticAnnotation := pod.Annotations[resticPodAnnotation]
+	if resticAnnotation == "" {
+		return []string{}
+	}
+	return strings.Split(pod.Annotations[resticPodAnnotation], ",")
+}
+
+func contains(slice []string, key string) bool {
+	for _, i := range slice {
+		if i == key {
+			return true
+		}
+	}
+	return false
+}
+
+func isPVCBackedUpByRestic(pvcNamespace, pvcName string, podClient corev1client.PodsGetter) (bool, error) {
+	pods, err := getPodsUsingPVC(pvcNamespace, pvcName, podClient)
+	if err != nil {
+		return false, errors.WithStack(err)
+	}
+
+	for _, p := range pods {
+		resticVols := getPodVolumesUsingRestic(p)
+		if len(resticVols) > 0 {
+			volName, err := getPodVolumeNameForPVC(p, pvcName)
+			if err != nil {
+				return false, err
+			}
+			if contains(resticVols, volName) {
+				return true, nil
+			}
+		}
+	}
+
+	return false, nil
+}
+
+func getVolumeSnapshotClassForStorageClass(provisioner string, snapshotGetter snapshotter.VolumeSnapshotClassesGetter) (*snapshotv1beta1api.VolumeSnapshotClass, error) {
+	for {
+		snapshotClasses, err := snapshotGetter.VolumeSnapshotClasses().List(metav1.ListOptions{})
+		if err != nil {
+			return nil, errors.Wrap(err, "error listing volumesnapshot classes")
+		}
+		for _, sc := range snapshotClasses.Items {
+			if sc.Driver == provisioner {
+				return &sc, nil
+			}
+		}
+		if snapshotClasses.Continue == "" {
+			break
+		}
+	}
+	return nil, errors.Errorf("failed to get volumesnapshotclass for provisioner %s", provisioner)
+}
+
+func getVolumeSnapshotContentForVolumeSnapshot(volSnapName string, snapshotContentsGetter snapshotter.VolumeSnapshotContentsGetter) (*snapshotv1beta1api.VolumeSnapshotContent, error) {
+	for {
+		snapshotContents, err := snapshotContentsGetter.VolumeSnapshotContents().List(metav1.ListOptions{})
```

Comment:
When would there be more than 1 `VolumeSnapshotContent` for a given `VolumeSnapshot`?

Reply:
I'm asking because usually the relationship is, as I understand, usually a 1:1 relationship, much like a PVC to a PV.

Reply:
isn't there a two-way binding here, such that we could just repeatedly `Get` the VolumeSnapshot and wait for the relevant binding field/ref on the `VolumeSnapshot` to be set, rather than having to list all VolumeSnapshotContents?

Reply:
Just confirmed from the backup yaml, the links are all 1-way.
`volumesnapshotcontents.snapshot`  has the name of `volumeSnapshotClass` and object ref to `VolumeSnapshot`
`volumesnapshots` just has the name of the `volumeSnapshotClass`

Reply:
I take that back.
The `volumesnapshots.snapshot.storage.k8s.io`  has a `boundVolumeSnapshotContentName` field with the `volumesnapshotcontent` name

Reply:
Yeah, the binding should definitely be 2 way.

Via [`VolumeSnapshot.Status.BoundSnapshotContentName`](https://github.com/kubernetes-csi/external-snapshotter/blob/master/pkg/apis/volumesnapshot/v1beta1/types.go#L115) on the backup side and [`VolumeSnapshot.Spec.Source.VolumeSnapshotContentName`](https://github.com/kubernetes-csi/external-snapshotter/blob/master/pkg/apis/volumesnapshot/v1beta1/types.go#L102) on the restore side.

The difference between the two will be that `BoundVolumeSnapshotContentName` will be created in response to a request for the CSI controllers to make a snapshot. `VolumeSnapshotContentName` will be used in the "import" case when a storage-side snapshot already exists and the user or some program is constructing a Kubernetes representation of it.

Reply:
We can simplify this code to get the `volumesnapshot` that we just created and read the `status.BOundVolumeSnapshotContentName`